src/lib.rs

```diff
@@ -143,36 +146,17 @@ impl Query {
         Ok(())
     }
 
-    // We pass in a &mut Pairs instead of returning as it allows to
-    // pre-size the object size and to reuse the object across
-    // multiple function calls.
-    pub fn keys(&self, keys: &[Key], pairs: &mut Pairs) -> Result<()> {
-        let result = unsafe {
-            ffi::rill_query_keys(self.query, keys.as_ptr(), keys.len(), pairs.pairs)
-        };
-        if result.is_null() { return err(); }
-
-        pairs.pairs = result;
-        Ok(())
+    pub fn all(&self, col: ffi::rill_col) -> Result<Pairs> {
+        let ptr = unsafe { ffi::rill_query_all(self.query, col)};
+        if !ptr.is_null() { Ok(Pairs{ pairs: ptr }) } else { err() }
     }
 
-    // \todo We only ever expect one val query during the lifetime of
-    // the process so it's worth considering just returning a pairs
-    // and not worrying too much about performance.
-    pub fn vals(&self, vals: &[Val], pairs: &mut Pairs) -> Result<()> {
-        let result = unsafe {
-            ffi::rill_query_vals(self.query, vals.as_ptr(), vals.len(), pairs.pairs)
-        };
+    pub fn vals(&self, vals: Vec<ffi::rill_val_t>, pairs: &mut Pairs) -> Result<()> {
```

Comment:
It would make more sense if `vals` had the type `&[ffi::rill_val_t]`.

Reply:
Also, it would be best if none of the `ffi` types were leaked in the interface. I believe I have an equivalent val type somewhere in the rill namespace.

Reply:
What I tried previously didn't work, but I'll try a workaround.